### PR TITLE
azure_rm_virtualmachine: better support for User Assigned and System Assigned identities

### LINKED ISSUE #1173: azure_rm_virtualmachine: Ability to modify/unset User Assigned and System Assigned identities on VMs
##### SUMMARY

Currently only System Assigned managed identities can be enabled in Virtual Machines and only while creating the VM (you can not assign the identity once the machine is provisioned). It would be great if both System Assigned and User Assigned identities could be configured using azure_rm_virtualmachine.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azure_rm_virtualmachine

##### ADDITIONAL INFORMATION

I have forked the collection and made a first approach of this feature, but it would be great if we could discuss here if my approach is correct. Some samples that work in my fork:

```yaml
- name: Set identities IDs to test. Identities ansible-test-vm-identity and ansible-test-vm-identity-2 have to be created previously
  set_fact:
    user_identity_1: "/subscriptions/{{ azure_subscription_id }}/resourcegroups/{{ resource_group }}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ansible-test-vm-identity"
    user_identity_2: "/subscriptions/{{ azure_subscription_id }}/resourcegroups/{{ resource_group }}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ansible-test-vm-identity-2"

- name: Append another UserAssigned
  azure_rm_virtualmachine:
    name: "{{ vm_name }}"
    resource_group: "{{ resource_group }}"
    vm_identity:
      type: "UserAssigned"
      user_assigned_identities:
        id: 
        - "{{ user_identity_2 }}"
        append: true
  register: vm_output

- name: Add SystemAssigned
  azure_rm_virtualmachine:
    name: "{{ vm_name }}"
    resource_group: "{{ resource_group }}"
    vm_identity:
      type: "SystemAssigned, UserAssigned"
      user_assigned_identities:
        append: true
  register: vm_output

- name: Force only one UserAssigned
  azure_rm_virtualmachine:
    name: "{{ vm_name }}"
    resource_group: "{{ resource_group }}"
    vm_identity:
      type: "SystemAssigned, UserAssigned"
      user_assigned_identities:
        id: 
        - "{{ user_identity_2 }}"
        append: false
  register: vm_output

- name: Remove UserAssigned
  azure_rm_virtualmachine:
    name: "{{ vm_name }}"
    resource_group: "{{ resource_group }}"
    vm_identity:
      type: "SystemAssigned"
  register: vm_output

- name: Disable SystemAssigned
  azure_rm_virtualmachine:
    name: "{{ vm_name }}"
    resource_group: "{{ resource_group }}"
    vm_identity:
      type: "None"
  register: vm_output
```

### PR DESCRIPTION
**Pull request by cpiment**
##### SUMMARY

Allows to set/unset User and System managed identities to Virtual Machines

Fixes #1173 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

azure_rm_virtualmachine

##### ADDITIONAL INFORMATION

Before
```yaml
- name: Create minimal VM with UserAssigned Identity 
  azure_rm_virtualmachine:
    allocated: no
    resource_group: "{{ resource_group }}"
    name: "{{ vm_name }}"
    vm_identity: 'SystemAssigned'
    admin_username: "testuser"
    ssh_password_enabled: false    
    public_ip_allocation_method: Disabled
    location: westeurope
    network_interface_names:
      - name: "{{ interface_name }}-identity"
        resource_group: "{{ resource_group }}"
    ssh_public_keys:
      - path: /home/testuser/.ssh/authorized_keys
        key_data: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDfoYlIV4lTPZTv7hXaVwQQuqBgGs4yeNRX0SPo2+HQt9u4X7IGwrtXc0nEUm6LfaCikMH58bOL8f20NTGz285kxdFHZRcBXtqmnMz2rXwhK9gwq5h1khc+GzHtdcJXsGA4y0xuaNcidcg04jxAlN/06fwb/VYwwWTVbypNC0gpGEpWckCNm8vlDlA55sU5et0SZ+J0RKVvEaweUOeNbFZqckGPA384imfeYlADppK/7eAxqfBVadVvZG8IJk4yvATgaIENIFj2cXxqu2mQ/Bp5Wr45uApvJsFXmi+v/nkiOEV1QpLOnEwAZo6EfFS4CCQtsymxJCl1PxdJ5LD4ZOtP xiuxi.sun@qq.com"
    vm_size: Standard_D4s_v3
    virtual_network: "{{ network_name }}-identity"
    image:
      offer: UbuntuServer
      publisher: Canonical
      sku: 16.04-LTS
      version: latest
```
Now
```yaml
- name: Create minimal VM with UserAssigned Identity 
  azure_rm_virtualmachine:
    allocated: no
    resource_group: "{{ resource_group }}"
    name: "{{ vm_name }}"
    vm_identity:
      type: UserAssigned
      user_assigned_identities:
        id:
        - "{{ user_identity_1 }}"
    admin_username: "testuser"
    ssh_password_enabled: false    
    public_ip_allocation_method: Disabled
    location: westeurope
    network_interface_names:
      - name: "{{ interface_name }}-identity"
        resource_group: "{{ resource_group }}"
    ssh_public_keys:
      - path: /home/testuser/.ssh/authorized_keys
        key_data: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDfoYlIV4lTPZTv7hXaVwQQuqBgGs4yeNRX0SPo2+HQt9u4X7IGwrtXc0nEUm6LfaCikMH58bOL8f20NTGz285kxdFHZRcBXtqmnMz2rXwhK9gwq5h1khc+GzHtdcJXsGA4y0xuaNcidcg04jxAlN/06fwb/VYwwWTVbypNC0gpGEpWckCNm8vlDlA55sU5et0SZ+J0RKVvEaweUOeNbFZqckGPA384imfeYlADppK/7eAxqfBVadVvZG8IJk4yvATgaIENIFj2cXxqu2mQ/Bp5Wr45uApvJsFXmi+v/nkiOEV1QpLOnEwAZo6EfFS4CCQtsymxJCl1PxdJ5LD4ZOtP xiuxi.sun@qq.com"
    vm_size: Standard_D4s_v3
    virtual_network: "{{ network_name }}-identity"
    image:
      offer: UbuntuServer
      publisher: Canonical
      sku: 16.04-LTS
      version: latest
```

### CODE REVIEW

File: plugins/modules/azure_rm_virtualmachine.py [hunk 1]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                                user_assigned_identities_dict = { **user_assigned_identities_dict, **current_user_assigned_identities_dict}
+                            # Save the identity
+                            vm_resource.identity = self.compute_models.VirtualMachineIdentity(
+                                type=self.vm_identity.get('type'),
+                                user_assigned_identities=user_assigned_identities_dict
+                            )
+                        # If there are no identities in 'id' and 'UserAssigned' in type
+                        elif 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities',{}).get('id',[])) == 0:
+                            # Fail if append is False
+                            if vm_identity_user_assigned_append == False:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                            if vm_identity_user_assigned_append is False:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 2]
```diff
@@ -1339,6 +1376,37 @@ def exec_module(self, **kwargs):
...
+                    # If type in module args different from type of vm_dict, update identities
+                    elif vm_dict['identity']['type'] != self.vm_identity.get('type'):
+                        update_vm_identity = True
+                    # If type in module args contains 'UserAssigned'
+                    elif 'UserAssigned' in self.vm_identity.get('type'):
+                        # Create sets with current user identities and module args identities
+                        new_managed_identities = set(self.vm_identity.get('user_assigned_identities',{}).get('id',[]))
+                        current_managed_identities = set(vm_dict['identity']['userAssignedIdentities'].keys())
+                        # If new identities have to be appended to VM
+                        if self.vm_identity.get('user_assigned_identities',{}).get('append',False) == True:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if self.vm_identity.get('user_assigned_identities',{}).get('append',False) is True:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 3]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
                     if self.license_type is not None:
                         vm_resource.license_type = self.license_type
 
+                    if self.vm_identity is not None:
+                        # If 'append' is set to True save current user assigned managed identities to use later
+                        if self.vm_identity.get('user_assigned_identities',{}) is not None and self.vm_identity.get('user_assigned_identities',{}).get('append', False) == True:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if self.vm_identity.get('user_assigned_identities',{}) is not None and self.vm_identity.get('user_assigned_identities',{}).get('append', False) is True:
```

**Comment by cpiment** [comment]
I have found out that the user managed identities can be created during the tests using `azure_rm_resource` module, so I have modified the tests and edited the PR to remove the prerrequisites for the testing to succeed.

**Comment by Fred-sun** [comment]
@cpiment Please! Thanks you very much!

**Comment by cpiment** [comment]
@Fred-sun I have just commited your changes, thank you so much!

File: plugins/modules/azure_rm_virtualmachine.py [hunk 4]
```diff
@@ -372,8 +372,36 @@
...
+                type: str
+            user_assigned_identities:
+                description:
+                    - User Assigned Managed Identities and its options
+                required: false
+                type: dict
+                suboptions:
+                    id:
+                        description:
+                            - List of the user assigned identities IDs associated to the VM
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]: trailing whitespace
```suggestion
                            - List of the user assigned identities IDs associated to the VM
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 5]
```diff
@@ -949,6 +977,15 @@ def extract_names_from_blob_uri(blob_uri, storage_suffix):
     disable_password_authentication=dict(type='bool')
 )
 
+user_assigned_identities_spec = dict(
+    id=dict(type='list',default=[]),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
    id=dict(type='list', default=[]),
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 6]
```diff
@@ -949,6 +977,15 @@ def extract_names_from_blob_uri(blob_uri, storage_suffix):
...
+    append=dict(type='bool',default=True)
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
    append=dict(type='bool', default=True)
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 7]
```diff
@@ -949,6 +977,15 @@ def extract_names_from_blob_uri(blob_uri, storage_suffix):
...
+)
+
+managed_identity_spec = dict(
+    type=dict(type='str',choices=['SystemAssigned','UserAssigned','SystemAssigned, UserAssigned','None'],required=True),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
    type=dict(type='str', choices=['SystemAssigned', 'UserAssigned', 'SystemAssigned, UserAssigned', 'None'], required=True),
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 8]
```diff
@@ -949,6 +977,15 @@ def extract_names_from_blob_uri(blob_uri, storage_suffix):
...
+    user_assigned_identities=dict(type='dict',options=user_assigned_identities_spec,default={}),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
    user_assigned_identities=dict(type='dict', options=user_assigned_identities_spec, default={}),
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 9]
```diff
@@ -949,6 +977,15 @@ def extract_names_from_blob_uri(blob_uri, storage_suffix):
...
+)
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py:988]: expected 2 blank lines!
```suggestion
)

```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 10]
```diff
@@ -1339,6 +1376,37 @@ def exec_module(self, **kwargs):
...
+                    # If type different to None, and VM has no current identities, update identities
+                    elif 'identity' not in vm_dict:
+                        update_vm_identity = True
+                    # If type in module args different from type of vm_dict, update identities
+                    elif vm_dict['identity']['type'] != self.vm_identity.get('type'):
+                        update_vm_identity = True
+                    # If type in module args contains 'UserAssigned'
+                    elif 'UserAssigned' in self.vm_identity.get('type'):
+                        # Create sets with current user identities and module args identities
+                        new_managed_identities = set(self.vm_identity.get('user_assigned_identities',{}).get('id',[]))
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        new_managed_identities = set(self.vm_identity.get('user_assigned_identities', {}).get('id', []))
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 11]
```diff
@@ -1339,6 +1376,37 @@ def exec_module(self, **kwargs):
...
+                        current_managed_identities = set(vm_dict['identity']['userAssignedIdentities'].keys())
+                        # If new identities have to be appended to VM
+                        if self.vm_identity.get('user_assigned_identities',{}).get('append',False) is True:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if self.vm_identity.get('user_assigned_identities', {}).get('append', False) is True:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 12]
```diff
@@ -1339,6 +1376,37 @@ def exec_module(self, **kwargs):
...
+                            # and the union of identities is longer
+                            if len(current_managed_identities) != len(new_managed_identities.union(current_managed_identities)):
+                                #update identities
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                                # update identities
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 13]
```diff
@@ -1574,7 +1642,22 @@ def exec_module(self, **kwargs):
                         vm_resource.license_type = self.license_type
 
                     if self.vm_identity:
-                        vm_resource.identity = self.compute_models.VirtualMachineIdentity(type=self.vm_identity)
+                        # If UserAssigned in module args and ids specified
+                        if 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities',{}).get('id',[])) != 0:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities', {}).get('id', [])) != 0:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 14]
```diff
@@ -1574,7 +1642,22 @@ def exec_module(self, **kwargs):
...
+                            user_assigned_identities_dict = { uami:dict() for uami in self.vm_identity.get('user_assigned_identities').get('id') }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                            user_assigned_identities_dict = {uami: dict() for uami in self.vm_identity.get('user_assigned_identities').get('id')}
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 15]
```diff
@@ -1574,7 +1642,22 @@ def exec_module(self, **kwargs):
...
+                            # Append identities to the model
+                            vm_resource.identity = self.compute_models.VirtualMachineIdentity(
+                                type=self.vm_identity.get('type'),
+                                user_assigned_identities=user_assigned_identities_dict
+                            )
+                        # If UserAssigned in module args, but ids are not specified
+                        elif 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities',{}).get('id',[])) == 0:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        elif 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities', {}).get('id', [])) == 0:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 16]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
                     if self.license_type is not None:
                         vm_resource.license_type = self.license_type
 
+                    if self.vm_identity is not None:
+                        # If 'append' is set to True save current user assigned managed identities to use later
+                        if self.vm_identity.get('user_assigned_identities',{}) is not None and self.vm_identity.get('user_assigned_identities',{}).get('append', False) is True:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if self.vm_identity.get('user_assigned_identities', {}) is not None and self.vm_identity.get('user_assigned_identities', {}).get('append', False) is True:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 17]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            if 'identity' in vm_dict and 'userAssignedIdentities' in vm_dict['identity']:
+                                current_user_assigned_identities_dict = { uami:dict() for uami in vm_dict['identity']['userAssignedIdentities'].keys() }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                                current_user_assigned_identities_dict = {uami: dict() for uami in vm_dict['identity']['userAssignedIdentities'].keys()}
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 18]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                                vm_identity_user_assigned_append = True
+                            else:
+                                # Nothing to append to
+                                vm_identity_user_assigned_append = False
+                        else:
+                            # 'append' is False or unset
+                            vm_identity_user_assigned_append = False
+                        # If there are identities in 'id' and 'UserAssigned' in type
+                        if 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities',{}).get('id',[])) != 0:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        if 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities', {}).get('id', [])) != 0:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 19]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            user_assigned_identities_dict = { uami:dict() for uami in self.vm_identity.get('user_assigned_identities').get('id') }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                            user_assigned_identities_dict = {uami: dict() for uami in self.vm_identity.get('user_assigned_identities').get('id')}
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 20]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            # If there are identities to append, merge the dicts
+                            if vm_identity_user_assigned_append:
+                                user_assigned_identities_dict = { **user_assigned_identities_dict, **current_user_assigned_identities_dict}
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                                user_assigned_identities_dict = {**user_assigned_identities_dict, **current_user_assigned_identities_dict}
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 21]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            # Save the identity
+                            vm_resource.identity = self.compute_models.VirtualMachineIdentity(
+                                type=self.vm_identity.get('type'),
+                                user_assigned_identities=user_assigned_identities_dict
+                            )
+                        # If there are no identities in 'id' and 'UserAssigned' in type
+                        elif 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities',{}).get('id',[])) == 0:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]:
```suggestion
                        elif 'UserAssigned' in self.vm_identity.get('type') and len(self.vm_identity.get('user_assigned_identities', {}).get('id', [])) == 0:
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 22]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            # Fail if append is False
+                            if vm_identity_user_assigned_append is False:
+                                self.fail("UserAssigned specified but no User Assigned IDs provided and no UserAssigned identities are currently assigned to the VM")
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]: Too long (165 characters)
```suggestion
                                self.fail("UserAssigned specified but no User Assigned IDs provided and no UserAssigned identities are currently assigned to the VM")
```

File: plugins/modules/azure_rm_virtualmachine.py [hunk 23]
```diff
@@ -1813,6 +1896,47 @@ def exec_module(self, **kwargs):
...
+                            # If append is true, user is changing from 'UserAssigned' to 'SystemAssigned, UserAssigned' and wants to keep current UserAssigned identities
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]: Too long (169 characters)
```suggestion
                            # If append is true, user is changing from 'UserAssigned' to 'SystemAssigned, UserAssigned' and wants to keep current UserAssigned identities
```

File: plugins/modules/azure_rm_virtualmachine.py (hunk 22 above)
Comment (cpiment) [reply plugins/modules/azure_rm_virtualmachine.py]: I have divided this line in two. Do you need me to rephrase the error message? If code reaches this point, the module arguments specify that a User Assigned Managed Identity has to be appended to the VM, but there aren't identities in the module arguments nor in the current VM configuration

File: plugins/modules/azure_rm_virtualmachine.py (hunk 23 above)
Comment (cpiment) [reply plugins/modules/azure_rm_virtualmachine.py]: I have divided this line in two

File: plugins/modules/azure_rm_virtualmachine.py [hunk 24]
```diff
@@ -1813,6 +1897,49 @@ def exec_module(self, **kwargs):
                     if self.license_type is not None:
                         vm_resource.license_type = self.license_type
 
+                    if self.vm_identity is not None:
+                        # If 'append' is set to True save current user assigned managed identities to use later
+                        if self.vm_identity.get('user_assigned_identities', {}) is not None and self.vm_identity.get('user_assigned_identities', {}).get('append', False) is True:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]: line too long (178 > 160 characters)
```suggestion
                        if self.vm_identity.get('user_assigned_identities', {}) is not None and self.vm_identity.get('user_assigned_identities', {}).get('append', False) is True:
```

Comment (cpiment) [reply plugins/modules/azure_rm_virtualmachine.py]: I have splitted this line in two

File: plugins/modules/azure_rm_virtualmachine.py [hunk 25]
```diff
@@ -1813,6 +1897,50 @@ def exec_module(self, **kwargs):
                     if self.license_type is not None:
                         vm_resource.license_type = self.license_type
 
+                    if self.vm_identity is not None:
+                        # If 'append' is set to True save current user assigned managed identities to use later
+                        if (self.vm_identity.get('user_assigned_identities', {}) is not None
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_virtualmachine.py]: Delete whitespace!
```suggestion
                        if (self.vm_identity.get('user_assigned_identities', {}) is not None
```

**Comment by cpiment** [comment]
Hi @Fred-sun 

Is there anything else to review?

Thanks!

**Comment by Fred-sun** [comment]
@cpiment  Perfact! I will push for merged! Thank you very much!

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
@cpiment LGTM

**Pull request merged**